src/apps/Permissions/CreatePermissionPanel.js

```diff
@@ -0,0 +1,299 @@
+import { PermissionsConsumer } from '../../contexts/PermissionsContext'
+import { ANY_ADDRESS } from '../../permissions'
+import { isAddress } from '../../web3-utils'
+import AppInstanceLabel from './AppInstanceLabel'
+
+const DEFAULT_STATE = {
+  assignEntityIndex: 0,
+  assignAddressValue: '',
+  appIndex: 0,
+  roleIndex: 0,
+}
+
+// The permission panel, wrapped in a PermissionsContext (see end of file)
+class PermissionPanel extends React.PureComponent {
+  state = {
+    ...DEFAULT_STATE,
+  }
+
+  handleAssignEntityChange = index => {
+    this.setState({ assignEntityIndex: index, assignAddressValue: '' })
+  }
+
+  handleAssignAddressChange = event => {
+    this.setState({ assignAddressValue: event.target.value })
+  }
+
+  handleAppChange = index => {
+    this.setState({ appIndex: index, roleIndex: 0 })
+  }
+
+  handleRoleChange = index => {
+    this.setState({ roleIndex: index })
+  }
+
+  getNamedApps() {
+    const { apps } = this.props
+    return apps.filter(app => Boolean(app.name))
+  }
+
+  getRoles() {
+    const { getAppRoles } = this.props
+    const app = this.getSelectedApp()
+    const appRoles = app ? getAppRoles(app) : []
+    return appRoles.map(({ role }) => role)
+  }
+
+  appsLabels() {
+    return this.getNamedApps().map(app => (
+      <AppInstanceLabel
+        app={app}
+        proxyAddress={app.proxyAddress}
+        coreRole={app.name === 'ACL' ? { appName: 'ACL' } : null}
+      />
+    ))
+  }
+
+  getAssignEntityItems() {
+    return [
+      'Select an entity',
+      ...this.appsLabels(),
+      'Any account',
+      'Custom address…',
+    ]
+  }
+
+  getAppsItems() {
+    return ['Select an app', ...this.appsLabels()]
+  }
+
+  getRolesItems() {
+    const roles = this.getRoles()
+    const names = roles.map(role => role.name || '?')
+    return ['Select a role', ...names]
+  }
+
+  getSelectedApp() {
+    // -1 for the “select an app” entry
+    return this.getNamedApps()[this.state.appIndex - 1]
+  }
+
+  // Get the address of the selected entity
+  getEntityAddress() {
+    const { assignEntityIndex, assignAddressValue } = this.state
+    const items = this.getAssignEntityItems()
+
+    // custom address
+    if (assignEntityIndex === items.length - 1) {
+      return assignAddressValue.trim()
+    }
+
+    // any account
+    if (assignEntityIndex === items.length - 2) {
+      return ANY_ADDRESS
+    }
+
+    // app proxy address (in the menu, apps have two entries above them)
+    const app = this.getNamedApps()[assignEntityIndex - 1]
+
+    const address = app ? app.proxyAddress : null
+    return address && address.trim()
+  }
+
+  canSubmit(assignEntityItems, rolesItems) {
+    const { assignEntityIndex, assignAddressValue, roleIndex } = this.state
+
+    // No entity selected
+    if (assignEntityIndex === 0) {
+      return false
+    }
+
+    // No custom address set
+    if (
+      assignEntityIndex === assignEntityItems.length - 1 &&
+      !isAddress(assignAddressValue.trim())
```

Comment:
If it is an address, we should also check if it's properly checksummed.

Reply:
`isAddress` does it already https://web3js.readthedocs.io/en/1.0/web3-utils.html#isaddress

Reply:
Ohh good point!